python/ml4ir/base/model/architectures/auto_dag_network.py

```diff
@@ -0,0 +1,362 @@
+        """
+        Define the layer operation for a layer in the model config
+
+        Parameters
+        ----------
+        layer_args: dict
+            All the arguments from model config which are needed to define a layer op
+
+        Returns
+        -------
+        dict
+            Dictionary with the layer instance, inputs required for the layer and the format of inputs to the layer
+        """
+        return {
+            self.INPUTS: layer_args[self.INPUTS],
+            self.OP_IDENTIFIER: self.instantiate_op(layer_args[self.LAYER_TYPE], layer_args.get(self.LAYER_KWARGS, {})),
+            self.INPUTS_AS_LIST: layer_args.get(self.INPUTS_AS_LIST, False)
+        }
+
+    def define_architecture(self, model_config: dict) -> LayerGraph:
+        """
+        Convert the model from model_config to a LayerGraph
+
+        Parameters
+        ----------
+        model_config: dict
+            Model config parsed into a dictionary
+
+        Returns
+        -------
+        LayerGraph
+            Dependency DAG for the given model config
+        """
+
+        layer_ops = {layer_args[self.NAME]: self.get_layer_op(layer_args) for layer_args in
+                     model_config[self.LAYERS_IDENTIFIER]}
+        # While sorting is not mandatory, I would advise not removing it for the sake of reproducibility
+        inputs = sorted(set([input_name for layer_op in layer_ops.values()
+                             for input_name in layer_op[self.INPUTS] if input_name not in layer_ops.keys()]))
+        return LayerGraph(layer_ops, inputs)
+
+    def call(self, inputs, training=None):
+        """
+        Perform the forward pass for the architecture layer
+
+        Parameters
+        ----------
+        inputs: dict of dict of tensors
+            Input feature tensors divided as train and metadata
+        training: bool
+            Boolean to indicate if the layer is used in training or inference mode
+
+        Returns
+        -------
+        tf.Tensor
+            Logits tensor computed with the forward pass of the architecture layer
+        """
+        train_features = inputs[FeatureTypeKey.TRAIN]
+
+        # Do not modify the input
+        outputs = {k: v for k, v in train_features.items()}
+
+        # Pass features through all the layers of the Model
+        for node in self.execution_order:
+            # Input nodes don't need any execution
+            if node.is_input_node:
+                if node.name not in train_features:
+                    raise KeyError(f"Input feature {node.name} cannot be found in the feature ops outputs")
+            else:
+                layer_input = {k: outputs[k] for k in node.inputs}
+                if node.inputs_as_list or len(layer_input) == 1:
+                    layer_input = list(layer_input.values())
+                    if len(layer_input) == 1:
+                        layer_input = layer_input[0]
```

Comment:
@arvindsrikantan Why would you need to do this? For `concat`?

Reply:
Not just for concat. This will be needed by all layers which have only a single tensor input or layers like concat which need list of inputs (e.g.: BERT - two segments)